server/models/dataset.py

```diff
@@ -294,17 +300,218 @@ def addImage(self, dataset, imageDataStream, imageDataSize, filename, signature,
+            's3',
+            aws_access_key_id=accessKeyId,
+            aws_secret_access_key=secretAccessKey)
+        try:
+            s3.delete_object(
+                Bucket=s3BucketName,
+                Key=s3ObjectKey
+            )
+        except botocore.exceptions.ClientError as e:
+            raise GirderException('Error deleting object: %s' % e.response['Error']['Message'])
+
+        Batch().remove(batch)
+
+        # TODO: Could update bucket policy to effectively revoke temporary security credentials
+
+    def finalizeZipUploadS3(self, dataset, batch, user, sendMail=False):
+        """
+        Finalize a direct-to-S3 upload of a ZIP file of images. This involves several
+        steps:
+        - Download the ZIP file from S3 and add it as a file attached to the dataset.
+        - Delete the ZIP file from S3.
+        - Ingest images from the ZIP file into the dataset.
+        """
+        # Get ZIP upload settings
+        accessKeyId = Setting().get(
+            PluginSettings.ZIP_UPLOAD_USER_ACCESS_KEY_ID)
+        secretAccessKey = Setting().get(
+            PluginSettings.ZIP_UPLOAD_USER_SECRET_ACCESS_KEY)
+        if not all([accessKeyId, secretAccessKey]):
+            raise GirderException('ZIP upload not configured.')
+
+        # Get upload information stored on batch
+        s3BucketName = batch.get('s3BucketName')
+        s3ObjectKey = batch.get('s3ObjectKey')
+        if not all([s3BucketName, s3ObjectKey]):
+            raise GirderException('Error retrieving upload information.')
+
+        # Move file from S3 to the assetstore, attached to the dataset
+        s3 = boto3.client(
+            's3',
+            aws_access_key_id=accessKeyId,
+            aws_secret_access_key=secretAccessKey)
+        try:
+            with TempDir() as tempDir:
+                # Download file from S3 as upload user
+                fileName = os.path.join(tempDir, str(batch['_id']) + '.zip')
+                s3.download_file(
+                    Bucket=s3BucketName,
+                    Key=s3ObjectKey,
+                    Filename=fileName
+                )
+
+                # Attach file to dataset
+                fileSize = os.path.getsize(fileName)
+                with open(fileName, 'rb') as fileStream:
+                    zipFile = Upload().uploadFromFile(
+                        obj=fileStream,
+                        size=fileSize,
+                        name=fileName,
+                        parentType='dataset',
+                        parent=dataset,
+                        attachParent=True,
+                        user=user,
+                        mimeType='application/zip'
+                    )
+                # TODO: remove this once a bug in upstream Girder is fixed
+                zipFile['attachedToType'] = ['dataset', 'isic_archive']
+                zipFile = File().save(zipFile)
+
+                # Delete file from S3 as upload user
+                s3.delete_object(
```

Comment:
Maybe we should reconsider this entire step, and start storing the ZIP files in S3 permanently (either within an S3 assetstore or as a more directly-referenced S3 object). We'd still need to download them once locally for extraction, obviously.